File: Objective-C/Sorting/sorting algorithms/BubbleSort.m
//
//  BubbleSort.m
//  Sorting
//
//  Created by Paul Kraft on 14.04.16.
//  Copyright © 2016 pauljohanneskraft. All rights reserved.
//

#import "BubbleSort.h"

@implementation BubbleSort

+ (void) sort:(NSMutableArray *)array fromLeft:(long)start toRight:(long)end {
    bool didChange;
    do {
        didChange = false;
        for(long i = start; i < end; i++) {
            if( [array objectAtIndex:i] > [array objectAtIndex:i+1] ) {
                [self swap:array index1:i index2:i+1];
                didChange = true;
            }
        }
    } while( didChange );
}

@end

File: Swift/Sort/sorting algorithms/BubbleSort.swift
//
//  BubbleSort.swift
//  Sort
//
//  Created by Paul Kraft on 21.12.15.
//  Copyright © 2015 Paul Kraft. All rights reserved.
//

func bubbleSort<T: Comparable>(var unsorted: [T]) -> (name: String, array: [T]) {
    var didChange = false
    repeat
    {
        didChange = false
        for i in 0..<unsorted.count - 1 {
            if unsorted[i+1] < unsorted[i] {
                didChange = true
                (unsorted[i], unsorted[i+1]) = (unsorted[i+1], unsorted[i])
            }
        }
    }
    while(didChange)
    return ("BubbleSort", unsorted)
}

func bubbleSortInPlace<T: Comparable>(inout unsorted: [T]) -> (name: String, array: [T]) {
    var didChange = false
    repeat
    {
        didChange = false
        for i in 0..<unsorted.count - 1 {
            if unsorted[i+1] < unsorted[i] {
                didChange = true
                (unsorted[i], unsorted[i+1]) = (unsorted[i+1], unsorted[i])
            }
        }
    }
        while(didChange)
    return ("BubbleSort", unsorted)
}